File: GO_ogle/MineSweeper.py
class Solution(object):
    def updateBoard(self, board, click):
        """
        :type board: List[List[str]]
        :type click: List[int]
        :rtype: List[List[str]]
        """
        x, y = click
        self.traverse(board, x, y)
        return board

    def traverse(self, board, x, y):
        """
        :type board: List[List[str]]
        :type x: int
        :type y: int
        :rtype: None
        """
        if x > len(board) - 1 or y > len(board[0]) - 1 or x < 0 or y < 0:
            return
        current_cell = board[x][y]
        if current_cell == 'M':
            board[x][y] = 'X'
        elif current_cell == 'X' or current_cell.isdigit():
            return
        elif self.diagonal_bomb(x, y, board) > 0:
            pass
        elif current_cell == 'E':
            board[x][y] = 'B'
            self.traverse(board, x + 1, y + 1)
            self.traverse(board, x - 1, y - 1)
            self.traverse(board, x - 1, y + 1)
            self.traverse(board, x + 1, y - 1)
            self.traverse(board, x + 1, y)
            self.traverse(board, x, y + 1)
            self.traverse(board, x - 1, y)
            self.traverse(board, x, y - 1)

    def diagonal_bomb(self, i, j, board):
        """
        :type i: int
        :type j: int
        :type board: List[List[int]]
        :rtype: None
        """

        top = [(-1, -1), (-1, 0), (-1, 1)]
        row = [(0, -1), (0, 1)]
        bottom = [(1, -1), (1, 0), (1, 1)]
        result_list = list()
        result_list.append(self.calculate_bombs(top, i, j, board))
        result_list.append((self.calculate_bombs(row, i, j, board)))
        result_list.append((self.calculate_bombs(bottom, i, j, board)))

        bombes_around = result_list[0] + result_list[1] + result_list[2]
        if bombes_around > 0:
            board[i][j] = str(bombes_around)

        return bombes_around

    def calculate_bombs(self, indexes, i, j, board):
        """
        :type indexes: List[int]
        :type i: int
        :type j: int
        :type board: List[List[int]]
        :rtype: int
        """
        sum_items = 0
        for index in indexes:
            if i + index[0] > -1 and j + index[1] > -1 and ((i + index[0]) < len(board)) and ((j + index[1]) < len(board[0])):
                    sum_items += 1 if board[i + index[0]][j + index[1]] == 'M' else 0
        return sum_items


solution = Solution()
print(solution.updateBoard([["E","M","M","E","E","E","E","E"],["E","E","M","E","E","E","E","E"],["E","E","E","E","E","E","E","E"],["E","M","E","E","E","E","E","E"],["E","E","E","E","E","E","E","E"],["E","E","M","E","E","E","E","E"],["E","E","E","E","E","E","E","E"],["E","E","E","E","E","E","E","E"]]
,[5,5]))

File: Uber/mineSweeper.scala
object Solution {
  private type Grid = Array[Array[Char]]
  private val MINE = 'M'
  private val REVEALED_MINE = 'X'
  private val EMPTY_SQUARE = 'E'
  private val REVEALED_EMPTY_SQUARE = 'B'
  private val DIRS =
    Array((1, 1), (-1, -1), (-1, 1), (1, -1), (1, 0), (0, 1), (-1, 0), (0, -1))
  private val TOP = Array((-1, -1), (-1, 0), (-1, 1))
  private val ROW = Array((0, -1), (0, 1))
  private val BOTTOM = Array((1, -1), (1, 0), (1, 1))
  private type Dirs = Array[(Int, Int)]
  def updateBoard(
      board: Array[Array[Char]],
      click: Array[Int]
  ): Array[Array[Char]] = {
    val Array(x, y) = click
    traverse(board, x, y)
    board
  }

  private def traverse(board: Grid, x: Int, y: Int): Unit = {
    if (!valid(board, x, y)) return
    val cell = board(x)(y)
    if (cell == MINE) {
      board(x)(y) = REVEALED_MINE
    } else if (cell == REVEALED_MINE || cell.isDigit) {
      return
    } else if (diagonalBomb(board, x, y) > 0) {
      return
    } else if (cell == EMPTY_SQUARE) {
      board(x)(y) = REVEALED_EMPTY_SQUARE
      for ((x1, y1) <- DIRS) {
        traverse(board, x + x1, y + y1)
      }
    }
  }

  private def diagonalBomb(board: Grid, x: Int, y: Int): Int = {
    val bombsAround = Array(TOP, ROW, BOTTOM).foldLeft(0) {
      case (bombs, dirs) => bombs + calculateBombs(dirs, board, x, y)
    }
    if (bombsAround > 0) {
      val bombsAsChar = s"${bombsAround}".toCharArray().head
      board(x)(y) = bombsAsChar
    }

    bombsAround
  }

  private def calculateBombs(dirs: Dirs, board: Grid, x: Int, y: Int): Int = {
    dirs.foldLeft(0) {
      case (count, (x1, y1)) => {
        val (xNxt, yNxt) = (x + x1, y + y1)
        if (valid(board, xNxt, yNxt)) {
          if (board(xNxt)(yNxt) == MINE) count + 1 else count
        } else count
      }
    }
  }

  private def valid(board: Grid, x: Int, y: Int): Boolean = {
    x > -1 && y > -1 && x < board.length && y < board.head.length
  }
}

/*
G: board: Array[Array[Char]] , click: Array[Int]
O: openedBoard: Array[Array[Char]]
T: O(nm)
S: O(nm)

Notes:
  - You are given a 2D char matrix representing the game board.
  - 'M' represents an unrevealed mine,
  - 'E' represents an unrevealed empty square,
  - 'B' represents a revealed blank square that has no adjacent (above, below, left, right, and all 4 diagonals) mines,
  - digit ('1' to '8') represents how many mines are adjacent to this revealed square,
  - and finally 'X' represents a revealed mine.
  - Rules:
    ^ If a mine ('M') is revealed, then the game is over - change it to 'X'.
    ^ If an empty square ('E') with no adjacent mines is revealed, then change it to revealed blank ('B') and all of its adjacent unrevealed squares should be revealed recursively.
    ^ If an empty square ('E') with at least one adjacent mine is revealed, then change it to a digit ('1' to '8') representing the number of adjacent mines.
    ^ Return the board when no more squares will be revealed.

Ex:

 */
